packages/store-api/src/platforms/vtex/resolvers/validateCart.ts

```diff
@@ -0,0 +1,162 @@
+import type { IStoreCart, IStoreOffer } from '../../../__generated__/schema'
+import type {
+  OrderForm,
+  OrderFormItem,
+  OrderFormInputItem,
+} from '../clients/commerce/types/OrderForm'
+import type { Context } from '..'
+
+const getId = (item: IStoreOffer) =>
+  [item.itemOffered.sku, item.seller.identifier, item.price].join('::')
+
+const orderFormItemToOffer = (
+  item: OrderFormItem,
+  index?: number
+): IStoreOffer & { index?: number } => ({
+  listPrice: item.listPrice / 100,
+  price: item.sellingPrice / 100,
+  quantity: item.quantity,
+  seller: { identifier: item.seller },
+  itemOffered: {
+    sku: item.id,
+    image: [],
+    name: item.name,
+  },
+  index,
+})
+
+const offerToOrderItemInput = (
+  offer: IStoreOffer & { index?: number }
+): OrderFormInputItem => ({
+  quantity: offer.quantity,
+  seller: offer.seller.identifier,
+  id: offer.itemOffered.sku,
+  index: offer.index,
+})
+
+const groupById = (offers: IStoreOffer[]): Map<string, IStoreOffer> =>
+  offers.reduce((acc, item) => {
+    const id = getId(item)
+
+    acc.set(id, acc.get(id) ?? item)
+
+    return acc
+  }, new Map<string, IStoreOffer>())
+
+const equals = (of1: OrderForm, of2: OrderForm) => {
+  const pick = ({ orderFormId, messages, items, salesChannel }: OrderForm) => ({
+    orderFormId,
+    messages,
+    salesChannel,
+    items: items.map(
+      ({ uniqueId, quantity, seller, sellingPrice, availability }) => ({
+        uniqueId,
+        quantity,
+        seller,
+        sellingPrice,
+        availability,
+      })
+    ),
+  })
+
+  return JSON.stringify(pick(of1)) === JSON.stringify(pick(of2))
+}
+
+/**
+ * This resolver implements the optimistic cart behavior. The main idea in here
+ * is that we receive a cart from the UI (as query params) and we validate it with
+ * the commerce platform. If the cart is valid, we return null, if the cart is
+ * invalid according to the commerce platform, we return the new cart the UI should use
+ * instead
+ *
+ * The algoritm is something like:
+ * 1. Fetch orderForm from VTEX
+ * 2. Compute delta changes between the orderForm and the UI's cart
+ * 3. Update the orderForm accordingly
```

Comment:
Why not just return the new the orderform on this step?

Reply:
We need to update the orderr form on the VTEX API